File: app/src/main/java/com/kent/university/privelt/utils/sentence/RuleEnum.java
package com.kent.university.privelt.utils.sentence;

import android.util.Log;
import lombok.AllArgsConstructor;

@AllArgsConstructor
public enum RuleEnum {
    REGULAR(new RuleReplacement[]{
            new RuleReplacement("", "s", true)
    }), SINGULAR(new RuleReplacement[]{
            new RuleReplacement("ss", "es", true),
            new RuleReplacement("s", "es", true),
            new RuleReplacement("h", "es", true),
            new RuleReplacement("x", "es", true),
            new RuleReplacement("z", "es", true),
            new RuleReplacement("y", "ies", true),
            new RuleReplacement("o", "oes", true),
            new RuleReplacement("is", "es", true),
    }), EXCEPTION(new RuleReplacement[]{
            new RuleReplacement("fez", "zes", true),
            new RuleReplacement("gas", "ses", true),
            new RuleReplacement("wife", "wives", false),
            new RuleReplacement("wolf", "wolves", false),
            new RuleReplacement("roof", "s", true),
            new RuleReplacement("belief", "s", true),
            new RuleReplacement("chef", "s", true),
            new RuleReplacement("ray", "s", true),
            new RuleReplacement("boy", "s", true),
            new RuleReplacement("photo", "s", true),
            new RuleReplacement("piano", "s", true),
            new RuleReplacement("halo", "s", true),
            new RuleReplacement("piano", "s", true),
            new RuleReplacement("phenomenon", "phenomena", false),
            new RuleReplacement("criterion", "criteria", false),
            new RuleReplacement("child", "ren", true),
            new RuleReplacement("goose", "geese", false),
            new RuleReplacement("man", "men", false),
            new RuleReplacement("woman", "women", false),
            new RuleReplacement("tooth", "teeth", false),
            new RuleReplacement("foot", "feet", false),
            new RuleReplacement("mouse", "mice", false),
            new RuleReplacement("person", "people", false),
    }), NO_CHANGES(new RuleReplacement[]{
            new RuleReplacement("sheep", "sheep", false),
            new RuleReplacement("series", "series", false),
            new RuleReplacement("species", "sheep", false),
            new RuleReplacement("deer", "deer", false),
    });

    private final RuleReplacement[] ruleReplacements;

    public static RuleReplacement findRuleReplacement(String word, RuleEnum ruleEnum) {
        for (RuleReplacement ruleReplacement : ruleEnum.ruleReplacements)
            if (word.endsWith(ruleReplacement.getEnding()))
                return ruleReplacement;
        return null;
    }

    public static RuleReplacement findRuleAdapter(String word) {
        RuleReplacement ruleReplacement = findRuleReplacement(word, NO_CHANGES);
        if (ruleReplacement != null) return ruleReplacement;

        ruleReplacement = findRuleReplacement(word, EXCEPTION);
        if (ruleReplacement != null) return ruleReplacement;

        ruleReplacement = findRuleReplacement(word, SINGULAR);
        if (ruleReplacement != null) return ruleReplacement;

        return REGULAR.ruleReplacements[0];
    }
}

File: app/src/main/java/com/kent/university/privelt/utils/sentence/RuleEnum.kt
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.kent.university.privelt.utils.sentence

import lombok.AllArgsConstructor

@AllArgsConstructor
enum class RuleEnum(private val ruleReplacements: Array<RuleReplacement>) {
    REGULAR(arrayOf<RuleReplacement>(
            RuleReplacement("", "s", true)
    )),
    SINGULAR(arrayOf<RuleReplacement>(
            RuleReplacement("ss", "es", true),
            RuleReplacement("s", "es", true),
            RuleReplacement("h", "es", true),
            RuleReplacement("x", "es", true),
            RuleReplacement("z", "es", true),
            RuleReplacement("y", "ies", true),
            RuleReplacement("o", "oes", true),
            RuleReplacement("is", "es", true))),
    EXCEPTION(arrayOf<RuleReplacement>(
            RuleReplacement("fez", "zes", true),
            RuleReplacement("gas", "ses", true),
            RuleReplacement("wife", "wives", false),
            RuleReplacement("wolf", "wolves", false),
            RuleReplacement("roof", "s", true),
            RuleReplacement("belief", "s", true),
            RuleReplacement("chef", "s", true),
            RuleReplacement("ray", "s", true),
            RuleReplacement("boy", "s", true),
            RuleReplacement("photo", "s", true),
            RuleReplacement("piano", "s", true),
            RuleReplacement("halo", "s", true),
            RuleReplacement("piano", "s", true),
            RuleReplacement("phenomenon", "phenomena", false),
            RuleReplacement("criterion", "criteria", false),
            RuleReplacement("child", "ren", true),
            RuleReplacement("goose", "geese", false),
            RuleReplacement("man", "men", false),
            RuleReplacement("woman", "women", false),
            RuleReplacement("tooth", "teeth", false),
            RuleReplacement("foot", "feet", false),
            RuleReplacement("mouse", "mice", false),
            RuleReplacement("person", "people", false))),
    NO_CHANGES(arrayOf<RuleReplacement>(
            RuleReplacement("sheep", "sheep", false),
            RuleReplacement("series", "series", false),
            RuleReplacement("species", "sheep", false),
            RuleReplacement("deer", "deer", false)));

    companion object {
        private fun findRuleReplacement(word: String, ruleEnum: RuleEnum): RuleReplacement? {
            for (ruleReplacement in ruleEnum.ruleReplacements) if (word.endsWith(ruleReplacement.ending)) return ruleReplacement
            return null
        }

        @JvmStatic
        fun findRuleAdapter(word: String): RuleReplacement {
            var ruleReplacement = findRuleReplacement(word, NO_CHANGES)
            if (ruleReplacement != null) return ruleReplacement
            ruleReplacement = findRuleReplacement(word, EXCEPTION)
            if (ruleReplacement != null) return ruleReplacement
            ruleReplacement = findRuleReplacement(word, SINGULAR)
            return ruleReplacement ?: REGULAR.ruleReplacements[0]
        }
    }

}